airbyte-integrations/postgres-destination/src/test/java/io/airbyte/integrations/destination/postgres/PostgresDestinationTest.java

```diff
@@ -0,0 +1,219 @@
+  }
+
+  @Test
+  void testWriteSuccess() throws Exception {
+    final DestinationConsumer<SingerMessage> consumer = new PostgresDestination().write(config, CATALOG);
+
+    consumer.accept(SINGER_MESSAGE_USERS1);
+    consumer.accept(SINGER_MESSAGE_TASKS1);
+    consumer.accept(SINGER_MESSAGE_USERS2);
+    consumer.accept(SINGER_MESSAGE_TASKS2);
+    consumer.accept(SINGER_MESSAGE_RECORD);
+    consumer.close();
+
+    Set<JsonNode> usersActual = recordRetriever(USERS_STREAM_NAME);
+    final Set<JsonNode> expectedUsersJson = Sets.newHashSet(SINGER_MESSAGE_USERS1.getRecord(), SINGER_MESSAGE_USERS2.getRecord());
+    assertEquals(expectedUsersJson, usersActual);
+
+    Set<JsonNode> tasksActual = recordRetriever(TASKS_STREAM_NAME);
+    final Set<JsonNode> expectedTasksJson = Sets.newHashSet(SINGER_MESSAGE_TASKS1.getRecord(), SINGER_MESSAGE_TASKS2.getRecord());
+    assertEquals(expectedTasksJson, tasksActual);
+
+    assertTmpTablesNotPresent(CATALOG.getStreams().stream().map(Stream::getName).collect(Collectors.toList()));
+  }
+
+  @SuppressWarnings("ResultOfMethodCallIgnored")
+  @Test
+  void testWriteFailure() throws Exception {
+    // hack to force an exception to be thrown from within the consumer.
+    final SingerMessage spiedMessage = spy(SINGER_MESSAGE_USERS1);
+    doThrow(new RuntimeException()).when(spiedMessage).getStream();
+
+    final DestinationConsumer<SingerMessage> consumer = spy(new PostgresDestination().write(config, CATALOG));
+
+    assertThrows(RuntimeException.class, () -> consumer.accept(spiedMessage));
+    consumer.accept(SINGER_MESSAGE_USERS2);
+    consumer.close();
+
+    final List<String> tableNames = CATALOG.getStreams().stream().map(Stream::getName).collect(toList());
+    assertTmpTablesNotPresent(CATALOG.getStreams().stream().map(Stream::getName).collect(Collectors.toList()));
+    // assert that no tables were created.
+    assertTrue(fetchNamesOfTablesInDb().stream().noneMatch(tableName -> tableNames.stream().anyMatch(tableName::startsWith)));
+  }
+
+  private List<String> fetchNamesOfTablesInDb() throws SQLException {
+    return DatabaseHelper.query(getDatabasePool(),
+        ctx -> ctx.fetch("SELECT table_name FROM information_schema.tables WHERE table_schema='public' AND table_type='BASE TABLE';"))
+        .stream()
+        .map(record -> (String) record.get("table_name")).collect(Collectors.toList());
+  }
+
+  private void assertTmpTablesNotPresent(List<String> tableNames) throws SQLException {
+    Set<String> tmpTableNamePrefixes = tableNames.stream().map(name -> name + "_").collect(Collectors.toSet());
+    assertTrue(fetchNamesOfTablesInDb().stream().noneMatch(tableName -> tmpTableNamePrefixes.stream().anyMatch(tableName::startsWith)));
+  }
+
+  private BasicDataSource getDatabasePool() {
+    return DatabaseHelper.getConnectionPool(db.getUsername(), db.getPassword(), db.getJdbcUrl());
+  }
+
+  private Set<JsonNode> recordRetriever(String streamName) throws Exception {
```

Comment:
this is duplicated from the integration test?

Reply:
yes.